The Coming Soon theme

MichaelPetrinolis opened the pull request:
based on the Building a theme for Orchard Core from scratch demo

sebastienros commented on src/OrchardCore/OrchardCore.Application.Cms.Targets/OrchardCore.Application.Cms.Targets.csproj:
```diff
@@ -12,6 +12,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\..\OrchardCore.Themes\TheComingSoonTheme\TheComingSoonTheme.csproj" />
```
Wrong location

sebastienros commented on src/OrchardCore.Themes/TheComingSoonTheme/Manifest.cs:
```diff
@@ -0,0 +1,9 @@
+    Author = "The Orchard Team",
+    Website = "http://orchardproject.net",
+    Version = "2.0.0",
+    Description = "A theme adapted for coming soon websites."
```
"Landing page for a project that is under construction"

sebastienros commented:
Should we provide a way to save the email addresses? A custom controller view or a razor page at least. I could imaging that we send an email to the one that is entered, probably by creating a custom user, sending a validation email.

MichaelPetrinolis commented on src/OrchardCore.Themes/TheComingSoonTheme/TheComingSoonTheme.csproj:
```diff
@@ -0,0 +1,30 @@
+  <ItemGroup>
+    <EmbeddedResource Remove="gulpfile.js;**\*.less;**\*.scss;" />
+    <Compile Remove="wwwroot\node_modules\**" />
+    <EmbeddedResource Remove="wwwroot\node_modules\**" />
```
If I understood I should remove node_modules and scss folders instead of ignoring *.scss

sebastienros replied:
We should never embed or add to source code the node modules.

MichaelPetrinolis replied:
I see your point, node_modules is already added to gitignore. So I removed the above block from csproj

Skrypt replied:
By the way one could want to keep deploying the SASS files for using Source Maps files.

jtkech commented:
Just for infos about module / theme project files.

We don't need to remove `node_modules` , it is done by `OC.Module.Targets.props`.

    <ModuleDefaultAssetExcludes>$(ModuleAssetExcludes);**\node_modules\**;node_modules\**</ModuleDefaultAssetExcludes>
    <ModuleDefaultAssetExcludes>$(ModuleDefaultAssetExcludes);**\jspm_packages\**;jspm_packages\**</ModuleDefaultAssetExcludes>
    <ModuleDefaultAssetExcludes>$(ModuleDefaultAssetExcludes);**\bower_components\**;bower_components\**</ModuleDefaultAssetExcludes>

That said the original pattern was to have a module `Assets.json` and put all gulp input assets `*.less`, `*.scss` and so on in the module `Assets` folder which is automatically excluded from embedded resources (see below), and run a gulp task to output final assets under `wwwroot` where all is embedded.

    <ModuleAssetExcludes>Assets.*;Assets\**;**\*.props;**\*.targets</ModuleAssetExcludes>

But, as you, `TheAgencyTheme` and `TheBlogTheme` have the gulpfile and all input assets under their `wwwroot` folder. So they (and you) have to explicitly exclude them from embedded resources, e.g by doing.

    <EmbeddedResource Remove="gulpfile.js;**\*.less;**\*.scss;" />

Finally, `TheAgencyTheme` and `TheBlogTheme` already use these kind of lines which are useless.

    <Compile Remove="wwwroot\node_modules\**" />
    <EmbeddedResource Remove="wwwroot\node_modules\**" />
    <None Remove="wwwroot\node_modules\**" />

So, because some themes put the gulpfile and all their gulp input assets under `wwwroot`, maybe we have to revise what is automatically excluded from embedded resources.

sebastienros commented:
We should not rely on the automatic stuff, it's just a safeguard. But explicitly do what we actually want: prevent all source assets (scss, less) from being embedded.

jtkech commented:
Okay.

So, if someone do a theme with `node_modules`, he has to pack it from a repo wich ignores it, as we do.

Skrypt commented:
We can alternatively generate Source Maps files for debugging SASS files which would me generated in the same folder as the CSS files.

MichaelPetrinolis commented:
my concept was to easily upgrade a theme by just copying the original assets under wwwroot, without needing to exclude/include custom assets, and by modifying/extending the gulp file to create a customized version for the theme. Maybe creating a theme_source directory where I can put the original theme files and with gulp tasks copy the output to wwwroot ?

MichaelPetrinolis commented:
@sebastienros I was thinking adding a marketing module with a basic mailing list management feature and use it from the theme.

sebastienros commented:
@MichaelPetrinolis can we do it in two steps? maybe three ;)

1- Get the theme merged
2- Get the form working
3- Mailing list module

You already did 1.
To do 2 we need a single controller/action to register the email entered. No email would be sent until the smtp settings are configured though, but that's fine, it just has to be documented. Or we extend Worflows to allow for user registration (and sending the confirmation email) which would be the same as the controller, but reusable for other sites (my preferred choice).
For 3 we need a new module, but I want to see what the features would be (export email? send emails?). You might need some other features we are already working on (Jobs Queue to queue mails).

MichaelPetrinolis commented:
I just saw the readme from https://github.com/OrchardCMS/OrchardCore/tree/dev/src/OrchardCore.Modules/OrchardCore.Forms (it is missing from documentation TOC). 
I will try to integrate Forms and Workflows in the Theme

agriffard commented:
Forms documenation is on the dev branch (last link of the 'CMS Modules' menu in TOC): https://orchardcore.readthedocs.io/en/dev/OrchardCore.Modules/OrchardCore.Forms/README/

MichaelPetrinolis commented:
@agriffard you are right, I was looking at https://orchardcore.readthedocs.io/en/latest/OrchardCore/OrchardCore.Modules/README/

sebastienros commented:
How is the progress?

MichaelPetrinolis commented:
The workflow does not start on POST, unless I regenerate the SAS token and update the ComingSoon content item. I think that we must create a new SAS token in the recipe every time it is run.

sebastienros commented:
The SAS token should only be required on GETs, the POST should only require an antiforgery token, maybe this is what you are missing.

MichaelPetrinolis commented:
The issue is in the following block in the Invoke Method of HttpWorkflowController
`if (!_securityTokenService.TryDecryptToken<WorkflowPayload>(token, out var payload))
 {
        _logger.LogWarning("Invalid SAS token provided");
        return NotFound();
 }`

sebastienros commented:
@sfmskywalker should there be an option on the activity in the case of a POST to not require a SAS?

sebastienros commented:
What I wrote is actually wrong. It looks like there is no js method to generate a token, so I assume you generated the token by hand, from a different site? I am writing a function for this, to be called from the recipe. I'll test it and let you know.

MichaelPetrinolis commented:
Thanx, if you point out an example I could add it to this PR :)

sebastienros commented:
Finally, a token is necessary to hide the workflowtype or id, but otherwise in this case we don't need a tokem, only an anti forgery token is necessary.

If we were to add an endpoint that accepts a workflowid and requires the antiforgery token only, one could forge a request and maybe trigger a specific workflow. I don't think this is an actual issue though.

But using a token at any times is also a valid solution, so here we can use a token that has an infinite lifetime. I will update this PR.

sebastienros commented:
I have pushed the change. 

What I noticed is that the workflow doesn't have a transition between the starting activity and the validation tasks, and when we submit it doesn't go back to the homepage without it.

Also the way `WorkflowExecutionContextHandlerBase` works, with a mix of `IGlobalMethodProvider` is not good. I wrongly assume I could use it as an `IGlobalMethodProvider` but I can't, I just created another class that contains the token generation.

MichaelPetrinolis commented:
Fixed the transition (it did not used the workflowactivityid variable). I added a js method to getRootUrl for a path, So it will work with tenants that have prefix instead of host. I selected the OrchardCore.Contents assembly to add it, if there is a better place please tell me to move it.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Contents/Scripting/UrlMethodsProvider.cs:
```diff
@@ -0,0 +1,36 @@
+        {
+            _getRootUrl = new GlobalMethod
+            {
+                Name = "getRootUrl",
```
`getUrlPrefix()` ?

I am adding so much value to this PR ;)

sebastienros replied:
I'll let you do this one if you agree with the change.

MichaelPetrinolis commented on src/OrchardCore.Modules/OrchardCore.Users/Workflows/Activities/RegisterUserTask.cs:
```diff
@@ -0,0 +1,80 @@
+                if (string.IsNullOrWhiteSpace(userName))
+                    userName = email;
+
+                IUser user = new User() { UserName = userName, Email = email };
```
made a minor change, it did not compile

MichaelPetrinolis commented on src/OrchardCore.Modules/OrchardCore.Workflows/Http/Startup.cs:
```diff
@@ -48,17 +49,26 @@ public override void ConfigureServices(IServiceCollection services)
+                name: "HttpWorkflow",
+                areaName: "OrchardCore.Workflows",
+                template: "workflows/{action}",
+                defaults: new { controller = nameof(HttpWorkflowController) }
```
I just tried it, it works ok

sebastienros commented:
Does the workflow actually sends an email to verify the user's email? Or is it part of a standard User registration setting?

MichaelPetrinolis commented:
it does not send an email, it just creates a user. Also it does not inform the initiator of the workflow on success. In order to work, one should configure nocaptcha or remove it from the workflow

sebastienros commented:
Having to configuring the recaptcha is fine. Bonus point if there is a field in the main content type that drives whether the recaptcha is necessary or not. But I think we should use the service that creates a user and sends a confirmation email. Otherwise the list would be filed with users that are either fake or wrong. The code already exists, we just need to set the flag, or use whatever settings are already set on the main User settings.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Users/Workflows/Activities/RegisterUserTask.cs:
```diff
@@ -45,7 +58,7 @@ public bool ConfirmUser
         public WorkflowExpression<string> ConfirmationEmailTemplate
         {
-            get => GetProperty(() => new WorkflowExpression<string>());
+            get => GetProperty(() => new WorkflowExpression<string>("Thank you for your interest! Please <a href=\"{{ emailConfirmationUrl }}\">click here</a> to confirm your email."));
```
This should be a property of the task itself. Even if the default is exactly this text.

MichaelPetrinolis replied:
removed the default text. Assign it from recipe and make clear in the editor form how to use Workflow.Properties.EmailConfirmationUrl

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Users/Workflows/Activities/RegisterUserTask.cs:
```diff
@@ -85,6 +103,35 @@ public override async Task<ActivityExecutionResult> ExecuteAsync(WorkflowExecuti
+                    var uriBuilder = new UriBuilder(request.Scheme, request.Host.Host);
+                    if (request.Host.Port.HasValue)
+                        uriBuilder.Port = request.Host.Port.Value;
+                    uriBuilder.Path = string.Format("OrchardCore.Users/Registration/ConfirmEmail?userId={0}&code={1}", user.Id, code);
```
Does the code already contain the user id?

MichaelPetrinolis replied:
The UserManager.ConfirmEmailAsync requires a User and a token. This is how it is done in RegistrationController.ConfirmEmail action.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Users/Workflows/Activities/RegisterUserTask.cs:
```diff
@@ -85,6 +103,35 @@ public override async Task<ActivityExecutionResult> ExecuteAsync(WorkflowExecuti
+                    var localized = new LocalizedHtmlString(nameof(RegisterUserTask), body);
+                    var message = new MailMessage() { Body = localized.IsResourceNotFound ? body : localized.Value, IsBodyHtml = true, BodyEncoding = Encoding.UTF8 };
+                    message.To.Add(email);
+                    var smtpService = (ISmtpService)_httpContextAccessor.HttpContext.RequestServices.GetService(typeof(ISmtpService));
```
Use `Microsoft.DependencyInjection.Extensions` and remove the cast.

sebastienros commented:
Looking nice, and close to the end, thanks again for this work.

MichaelPetrinolis commented:
We thank all the orchard team for their great work ! 🥇

sebastienros commented:
Are you done with the changes and waiting for merge?

MichaelPetrinolis commented:
I think there is an issue with the default tenant. Let me check it and signal when i think is ok for merge.

MichaelPetrinolis commented:
It seems ok. There is no thank you page, it does not exist in the original theme also.

sebastienros commented:
I will test it soon

sebastienros commented:
I merged dev, updated the source files from Start Bootstrap, and tweaked the recipe. Can you check the changes and if you are fine with these then I will merge.

MichaelPetrinolis commented:
I am ok with the changes, it seems to work ok.
I fixed an issue with the url in the email not working.
I added the Users.Registration feature in the recipe
I added a message on successful registration

There is an issue with the Users.Registration controller, ConfirmEmailAsync does
not validate the code.

sebastienros commented:
> There is an issue with the Users.Registration controller, ConfirmEmailAsync does not validate the code.

Do you mean that if you provide a wrong code it still works?

MichaelPetrinolis commented:
I checked with Allow Users registration (from login UI) and with the Register User Task Activity, both fail with the code token that is created.

sebastienros commented:
Maybe some encoding issue. I noticed the query string parameter was not url encoded. I will try.

sebastienros commented:
I fixed the registration controller. Now I will test the activity.

sebastienros commented:
It also works with the registration task. I just set it so that if the validation fails, it doesn't keep the workflow instance alive.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Users/Workflows/Activities/RegisterUserTask.cs:
```diff
@@ -0,0 +1,151 @@
+
+
+                    var body = await _expressionEvaluator.EvaluateAsync(ConfirmationEmailTemplate, workflowContext);
+                    var localized = new LocalizedHtmlString(nameof(RegisterUserTask), body);
+                    var message = new MailMessage() { Body = localized.IsResourceNotFound ? body : localized.Value, IsBodyHtml = true, BodyEncoding = Encoding.UTF8 };
```
@MichaelPetrinolis Do you remember or understand what you were trying to do here with regards to localization? I don't understand what it would do better than just assign body directly?
/cc @hishamco if you have any idea also.
I could think that you are trying to allow for localizing the body, but I don't think it would work.

hishamco replied:
Basically he checking the existing of the resource `RegisterUserTask` and set the body, but I don't understand why you need to do this check while the localization APIs did it for you moreover I don't think there a value of creating a localizer here, so I don't think it world work too ;)

hishamco replied:
I'm asking myself how you checked a PR since merged few years ago ;)

MichaelPetrinolis replied:
To tell the truth, it was one of my first PRs to learn OrchardCore. Watched the 'build a theme from scratch' demo and tried it my self :) Don't remember what drive me to this decision at that time.

MichaelPetrinolis replied:
I can send a PR to remove it in order to avoid the confusion

sebastienros replied:
I already have it covered, thanks, found other places I want to clean up.

hishamco replied:
I just want to clean it up, but while you already started it's in your plate
